hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCleanerChore.java

```diff
@@ -0,0 +1,289 @@
+    justification="Intentional")
+
+  protected void chore() {
+    try {
+
+      TableDescriptors htds = master.getTableDescriptors();
+      Map<String, TableDescriptor> map = htds.getAll();
+      for (TableDescriptor htd : map.values()) {
+        for (ColumnFamilyDescriptor hcd : htd.getColumnFamilies()) {
+          if (hcd.isMobEnabled() && hcd.getMinVersions() == 0) {
+            cleaner.cleanExpiredMobFiles(htd.getTableName().getNameAsString(), hcd);
+          }
+        }
+        // Now clean obsolete files for a table
+        LOG.info("Cleaning obsolete MOB files ...");
+        cleanupObsoleteMobFiles(master.getConfiguration(), htd.getTableName());
+        LOG.info("Cleaning obsolete MOB files finished");
+      }
+    } catch (Exception e) {
+      LOG.error("Fail to clean the expired mob files", e);
+    }
+  }
+  /**
+   * Performs housekeeping file cleaning (called by MOB Cleaner chore)
+   * @param conf configuration
+   * @param table table name
+   * @throws IOException
+   */
+  public void cleanupObsoleteMobFiles(Configuration conf, TableName table)
+      throws IOException {
+
+    this.minAgeToArchive = conf.getLong(MobConstants.MIN_AGE_TO_ARCHIVE_KEY,
+      MobConstants.DEFAULT_MIN_AGE_TO_ARCHIVE);
+    try (final Connection conn = ConnectionFactory.createConnection(conf);
+        final Admin admin = conn.getAdmin();) {
+      TableDescriptor htd = admin.getDescriptor(table);
+      List<ColumnFamilyDescriptor> list = MobUtils.getMobColumnFamilies(htd);
+      if (list.size() == 0) {
+        LOG.info("Skipping non-MOB table [{}]",  table);
+        return;
+      }
+      // We check only those MOB files, which creation time is less
+      // than maxTimeToArchive. This is a current time - 1h. 1 hour gap
+      // gives us full confidence that all corresponding store files will
+      // exist at the time cleaning procedure begins and will be examined.
+      // So, if MOB file creation time is greater than this maxTimeToArchive,
+      // this will be skipped and won't be archived.
+      long maxCreationTimeToArchive = EnvironmentEdgeManager.currentTime() - minAgeToArchive;
+      LOG.info("Only MOB files whose creation time less than {} will be archived",
+        maxCreationTimeToArchive);
+      Path rootDir = FSUtils.getRootDir(conf);
+      Path tableDir = FSUtils.getTableDir(rootDir, table);
+      // How safe is this call?
+      List<Path> regionDirs = FSUtils.getRegionDirs(FileSystem.get(conf), tableDir);
+
+      Set<String> allActiveMobFileName = new HashSet<String>();
+      FileSystem fs = FileSystem.get(conf);
+      for (Path regionPath : regionDirs) {
+        for (ColumnFamilyDescriptor hcd : list) {
+          String family = hcd.getNameAsString();
+          Path storePath = new Path(regionPath, family);
+          boolean succeed = false;
+          Set<String> regionMobs = new HashSet<String>();
+          while (!succeed) {
```

Comment:
what happens if someone e.g. deletes the table or removes the column family while we're in this loop? what happens if regions are merged or split while we're in here?

Reply:
Added the following code:
while (!succeed) {
            if (!fs.exists(storePath)) {
              LOG.warn("Directory {} was deleted during cleaner procedure execution," 
                  +" skipping.", storePath);
              break;
            }
            RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(storePath);
            List<Path> storeFiles = new ArrayList<Path>();

In the beginning of the loop I check if region directory exists, if doesn't  - break out of loop
If cf or table was deleted during run - we collect refs only partially and they will be preserved, but in any case, those files must be deleted by delete table/cf procedure

Splits or merges are fine, because, at any given time, either daughter regions or parent regions are present physically in a file system and we need either of them to get full list of references

If compaction happens and files get removed during cleaner run we will get FNFE and will resume this run from the beginning for a region.